File: tests/cpp/test_audio_validation.cpp
/**
 * Audio System Validation Tests
 * 
 * Tests input validation, error handling, and exception types.
 */

#include "test_common.h"
#include "audio_manager.h"
#include "audio_session.h"
#include <string>

using namespace audio;
using namespace std::chrono_literals;

void test_not_initialized() {
    TEST("Not Initialized Errors")
    
    AudioManager& manager = AudioManager::GetInstance();
    manager.Shutdown();  // Ensure system is not running
    
    ASSERT_THROWS(NotInitializedException,
        manager.SetMasterVolume(0.5f),
        "SetMasterVolume without Initialize throws NotInitializedException");
    
    ASSERT_THROWS(NotInitializedException,
        manager.CreateGroup("test_group"),
        "CreateGroup without Initialize throws NotInitializedException");
    
    ASSERT_THROWS(NotInitializedException,
        manager.CreateTrack(),
        "CreateTrack without Initialize throws NotInitializedException");
    
    ASSERT_THROWS(NotInitializedException,
        manager.LoadSound("some_file.wav"),
        "LoadSound without Initialize throws NotInitializedException");
    
    ASSERT_NO_THROW(
        manager.Initialize(),
        "Reinitialize after not initialized does not throw");
    
    END_TEST
}

void test_audio_session_lifecycle() {
    TEST("AudioSession Lifecycle")

    AudioManager& manager = AudioManager::GetInstance();
    manager.Shutdown();  // Ensure clean state

    AudioSession session;

    ASSERT_NO_THROW(
        manager.SetMasterVolume(0.5f),
        "AudioSession initializes audio system");

    session.Close();

    ASSERT_THROWS(NotInitializedException,
        manager.SetMasterVolume(0.5f),
        "AudioSession.Close shuts down when it owns initialization");

    ASSERT_NO_THROW(
        manager.Initialize(),
        "Reinitialize after AudioSession.Close does not throw");

    END_TEST
}

void test_invalid_track_handle() {
    TEST("Invalid Track Handle Operations")
    
    AudioManager& manager = AudioManager::GetInstance();
    
    // Test with completely invalid handle (0)
    ASSERT_THROWS(InvalidHandleException,
        manager.PlayTrack(TrackHandle{0}),
        "PlayTrack with handle 0 throws InvalidHandleException");
    
    ASSERT_THROWS(InvalidHandleException,
        manager.StopTrack(TrackHandle{0}),
        "StopTrack with handle 0 throws InvalidHandleException");
    
    ASSERT_THROWS(InvalidHandleException,
        manager.AddLayer(TrackHandle{0}, "layer1", sound_dir + "/digital_base.wav"),
        "AddLayer with handle 0 throws InvalidHandleException");
    
    ASSERT_THROWS(InvalidHandleException,
        manager.RemoveLayer(TrackHandle{0}, "layer1"),
        "RemoveLayer with handle 0 throws InvalidHandleException");
    
    ASSERT_THROWS(InvalidHandleException,
        manager.SetLayerVolume(TrackHandle{0}, "layer1", 0.5f),
        "SetLayerVolume with handle 0 throws InvalidHandleException");
    
    // Test with non-existent handle (999999)
    ASSERT_THROWS(InvalidHandleException,
        manager.PlayTrack(TrackHandle{999999}),
        "PlayTrack with handle 999999 throws InvalidHandleException");
    
    ASSERT_THROWS(InvalidHandleException,
        manager.StopTrack(TrackHandle{999999}),
        "StopTrack with handle 999999 throws InvalidHandleException");
    
    END_TEST
}

void test_invalid_sound_handle() {
    TEST("Invalid Sound Handle Operations")
    
    AudioManager& manager = AudioManager::GetInstance();
    
    // Test with invalid handle (0)
    ASSERT_THROWS(InvalidHandleException,
        manager.PlaySound(SoundHandle{0}),
        "PlaySound with handle 0 throws InvalidHandleException");
    
    ASSERT_THROWS(InvalidHandleException,
        manager.StopSound(SoundHandle{0}),
        "StopSound with handle 0 throws InvalidHandleException");
    
    ASSERT_THROWS(InvalidHandleException,
        manager.SetSoundVolume(SoundHandle{0}, 0.5f),
        "SetSoundVolume with handle 0 throws InvalidHandleException");
    
    // Test with non-existent handle (999999)
    ASSERT_THROWS(InvalidHandleException,
        manager.PlaySound(SoundHandle{999999}),
        "PlaySound with handle 999999 throws InvalidHandleException");
    
    ASSERT_THROWS(InvalidHandleException,
        manager.StopSound(SoundHandle{999999}),
        "StopSound with handle 999999 throws InvalidHandleException");
    
    END_TEST
}

void test_invalid_group_handle() {
    TEST("Invalid Group Handle Operations")
    
    AudioManager& manager = AudioManager::GetInstance();
    
    // Test with invalid handle (0)
    ASSERT_THROWS(InvalidHandleException,
        manager.SetGroupVolume(GroupHandle{0}, 0.5f),
        "SetGroupVolume with handle 0 throws InvalidHandleException");
    
    // Test with non-existent handle (999999)
    ASSERT_THROWS(InvalidHandleException,
        manager.SetGroupVolume(GroupHandle{999999}, 0.5f),
        "SetGroupVolume with handle 999999 throws InvalidHandleException");
    
    END_TEST
}

void test_file_not_found() {
    TEST("File Not Found Errors")
    
    AudioManager& manager = AudioManager::GetInstance();
    
    // Test loading non-existent sound file
    ASSERT_THROWS(FileLoadException,
        manager.LoadSound("nonexistent_file.mp3"),
        "LoadSound with non-existent file throws FileLoadException");
    
    ASSERT_THROWS(FileLoadException,
        manager.LoadSound("does_not_exist.wav"),
        "LoadSound with invalid path throws FileLoadException");
    
    // Test loading with empty filename
    ASSERT_THROWS(FileLoadException,
        manager.LoadSound(""),
        "LoadSound with empty filename throws FileLoadException");
    
    // Create a track first, then try to add invalid layer
    TrackHandle track = manager.CreateTrack();
    
    ASSERT_THROWS(FileLoadException,
        manager.AddLayer(track, "layer1", "nonexistent_layer.mp3"),
        "AddLayer with non-existent file throws FileLoadException");
    
    manager.DestroyTrack(track);
    
    END_TEST
}

void test_fade_duration_validation() {
    TEST("Fade Duration Validation")
    
    AudioManager& manager = AudioManager::GetInstance();
    
    GroupHandle group = manager.CreateGroup("test");
    TrackHandle track = manager.CreateTrack();
    manager.AddLayer(track, "layer1", sound_dir + "/digital_base.wav");
    
    // Test negative duration
    ASSERT_THROWS(AudioException,
        manager.FadeGroup(group, 0.5f, std::chrono::milliseconds(-100)),
        "FadeGroup with negative duration throws AudioException");
    
    ASSERT_THROWS(AudioException,
        manager.FadeLayer(track, "layer1", 0.5f, std::chrono::milliseconds(-50)),
        "FadeLayer with negative duration throws AudioException");
    
    // Test zero duration
    ASSERT_THROWS(AudioException,
        manager.FadeGroup(group, 0.5f, std::chrono::milliseconds(0)),
        "FadeGroup with zero duration throws AudioException");
    
    ASSERT_THROWS(AudioException,
        manager.FadeLayer(track, "layer1", 0.5f, std::chrono::milliseconds(0)),
        "FadeLayer with zero duration throws AudioException");
    
    // Cleanup
    manager.DestroyTrack(track);
    manager.DestroyGroup(group);
    
    END_TEST
}

void test_input_validation() {
    TEST("Input Validation for Layer Names and Paths")
    
    AudioManager& manager = AudioManager::GetInstance();
    
    TrackHandle track = manager.CreateTrack();
    
    // Test empty layer name
    ASSERT_THROWS(AudioException,
        manager.AddLayer(track, "", sound_dir + "/digital_base.wav"),
        "AddLayer with empty layer name throws AudioException");
    
    // Test empty filepath
    ASSERT_THROWS(AudioException,
        manager.AddLayer(track, "layer1", ""),
        "AddLayer with empty filepath throws AudioException");
    
    // Test empty folder path
    ASSERT_THROWS(AudioException,
        manager.PlayRandomSoundFromFolder("", GroupHandle::Invalid()),
        "PlayRandomSoundFromFolder with empty path throws AudioException");
    
    // Cleanup
    manager.DestroyTrack(track);
    
    END_TEST
}

void test_exception_messages() {
    TEST("Exception Messages Are Descriptive")
    
    AudioManager& manager = AudioManager::GetInstance();
    
    // Check that exception messages contain useful information
    try {
        manager.PlayTrack(TrackHandle{0});
        std::cerr << "  FAIL: Expected InvalidHandleException" << std::endl;
        tests_failed++;
    } catch (const InvalidHandleException& e) {
        std::string msg = e.what();
        if (msg.find("Invalid") != std::string::npos || 
            msg.find("handle") != std::string::npos) {
            std::cout << "  PASS: InvalidHandleException message is descriptive: " << msg << std::endl;
            tests_passed++;
        } else {
            std::cerr << "  FAIL: Exception message not descriptive: " << msg << std::endl;
            tests_failed++;
        }
    }
    
    try {
        manager.LoadSound("nonexistent.mp3");
        std::cerr << "  FAIL: Expected FileLoadException" << std::endl;
        tests_failed++;
    } catch (const FileLoadException& e) {
        std::string msg = e.what();
        if (msg.find("file") != std::string::npos || 
            msg.find("load") != std::string::npos ||
            msg.find("not found") != std::string::npos) {
            std::cout << "  PASS: FileLoadException message is descriptive: " << msg << std::endl;
            tests_passed++;
        } else {
            std::cerr << "  FAIL: Exception message not descriptive: " << msg << std::endl;
            tests_failed++;
        }
    }
    
    END_TEST
}

void test_exception_types_hierarchy() {
    TEST("Exception Type Hierarchy")
    
    AudioManager& manager = AudioManager::GetInstance();
    
    // Test that specific exceptions can be caught as AudioException
    try {
        manager.PlayTrack(TrackHandle{0});
        std::cerr << "  FAIL: Expected exception to be thrown" << std::endl;
        tests_failed++;
    } catch (const AudioException& e) {
        std::cout << "  PASS: InvalidHandleException caught as AudioException base class" << std::endl;
        tests_passed++;
    }
    
    try {
        manager.LoadSound("nonexistent.mp3");
        std::cerr << "  FAIL: Expected exception to be thrown" << std::endl;
        tests_failed++;
    } catch (const AudioException& e) {
        std::cout << "  PASS: FileLoadException caught as AudioException base class" << std::endl;
        tests_passed++;
    }
    
    END_TEST
}

void test_valid_operations_no_exceptions() {
    TEST("Valid Operations Should Not Throw")
    
    AudioManager& manager = AudioManager::GetInstance();
    
    // Load a valid sound
    SoundHandle sound_handle = SoundHandle::Invalid();
    ASSERT_NO_THROW(
        sound_handle = manager.LoadSound(sound_dir + "/digital_base.wav"),
        "LoadSound with valid file does not throw");
    
    // Start and stop the sound
    ASSERT_NO_THROW(
        manager.PlaySound(sound_handle),
        "PlaySound with valid handle does not throw");
    
    ASSERT_NO_THROW(
        manager.StopSound(sound_handle),
        "StopSound with valid handle does not throw");
    
    // Create and manipulate track
    TrackHandle track = TrackHandle::Invalid();
    ASSERT_NO_THROW(
        track = manager.CreateTrack(),
        "CreateTrack does not throw");
    
    ASSERT_NO_THROW(
        manager.AddLayer(track, "layer1", sound_dir + "/digital_base.wav"),
        "AddLayer with valid file does not throw");
    
    ASSERT_NO_THROW(
        manager.PlayTrack(track),
        "PlayTrack with valid handle does not throw");
    
    ASSERT_NO_THROW(
        manager.StopTrack(track),
        "StopTrack with valid handle does not throw");
    
    // Create and manipulate group
    GroupHandle group = GroupHandle::Invalid();
    ASSERT_NO_THROW(
        group = manager.CreateGroup("test_group"),
        "CreateGroup does not throw");
    
    ASSERT_NO_THROW(
        manager.SetGroupVolume(group, 0.5f),
        "SetGroupVolume with valid handle does not throw");
    
    // Cleanup
    manager.DestroySound(sound_handle);
    manager.DestroyTrack(track);
    manager.DestroyGroup(group);
    
    END_TEST
}

void test_multiple_invalid_operations() {
    TEST("Multiple Invalid Operations In Sequence")
    
    AudioManager& manager = AudioManager::GetInstance();
    
    // Multiple invalid track operations should all throw
    for (int i = 0; i < 3; i++) {
        ASSERT_THROWS(InvalidHandleException,
            manager.PlayTrack(TrackHandle{0}),
            "Multiple PlayTrack calls with invalid handle all throw");
    }
    
    // Multiple invalid file loads should all throw
    for (int i = 0; i < 3; i++) {
        ASSERT_THROWS(FileLoadException,
            manager.LoadSound("nonexistent.mp3"),
            "Multiple LoadSound calls with invalid file all throw");
    }
    
    END_TEST
}

int main(int argc, char* argv[]) {
    std::cout << "========================================" << std::endl;
    std::cout << "Audio System Validation Tests" << std::endl;
    std::cout << "========================================" << std::endl;
    
    sound_dir = get_sound_dir(argc, argv);
    std::cout << "Using sound files from: " << sound_dir << std::endl;
    
    // Initialize audio system
    try {
        AudioManager& manager = AudioManager::GetInstance();
        manager.Initialize();
        std::cout << "Audio system initialized" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Failed to initialize audio system: " << e.what() << std::endl;
        return 1;
    }
    
    // Run all error handling tests
    test_not_initialized();
    test_audio_session_lifecycle();
    test_invalid_track_handle();
    test_invalid_sound_handle();
    test_invalid_group_handle();
    test_file_not_found();
    test_fade_duration_validation();
    test_input_validation();
    test_exception_messages();
    test_exception_types_hierarchy();
    test_valid_operations_no_exceptions();
    test_multiple_invalid_operations();
    
    // Shutdown
    AudioManager& manager = AudioManager::GetInstance();
    manager.Shutdown();
    
    // Print results
    std::cout << "\n========================================" << std::endl;
    std::cout << "Test Results" << std::endl;
    std::cout << "========================================" << std::endl;
    std::cout << "Passed: " << tests_passed << std::endl;
    std::cout << "Failed: " << tests_failed << std::endl;
    
    if (tests_failed == 0) {
        std::cout << "\nAll validation tests passed!" << std::endl;
        return 0;
    } else {
        std::cout << "\nSome tests failed!" << std::endl;
        return 1;
    }
}

File: tests/python/test_audio_validation.py
"""
Audio System Validation Tests
Tests input validation, error handling, and exception types.
"""

from test_common import *

def test_not_initialized():
    """Test: Using API without initialize should raise NotInitializedException"""
    print("Test: Not initialized... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    audio.shutdown()
    
    try:
        audio.set_master_volume(0.5)
        print("FAIL - No exception raised")
        return False
    except game_audio.NotInitializedException:
        print("PASS")
        return True
    except Exception as e:
        print(f"FAIL - Wrong exception type: {type(e).__name__}: {e}")
        return False

def test_audio_session_lifecycle():
    """Test: AudioSession initializes and shuts down properly"""
    print("Test: AudioSession lifecycle... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    audio.shutdown()

    session = game_audio.AudioSession()
    try:
        audio.set_master_volume(0.5)
    except Exception as e:
        print(f"FAIL - Unexpected exception after session init: {type(e).__name__}: {e}")
        session.close()
        return False

    session.close()

    try:
        audio.set_master_volume(0.5)
        print("FAIL - No exception after session close")
        return False
    except game_audio.NotInitializedException:
        print("PASS")
        return True
    except Exception as e:
        print(f"FAIL - Wrong exception type: {type(e).__name__}: {e}")
        return False

def test_invalid_track_handle():
    """Test: Using invalid track handle should raise InvalidHandleException"""
    print("Test: Invalid track handle... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    audio.initialize()
    
    try:
        audio.play_track(game_audio.TrackHandle(99999))
        print("FAIL - No exception raised")
        audio.shutdown()
        return False
    except game_audio.InvalidHandleException:
        print("PASS")
        audio.shutdown()
        return True
    except Exception as e:
        print(f"FAIL - Wrong exception type: {type(e).__name__}: {e}")
        audio.shutdown()
        return False

def test_invalid_sound_handle():
    """Test: Using invalid sound handle should raise InvalidHandleException"""
    print("Test: Invalid sound handle... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    audio.initialize()
    
    try:
        audio.play_sound(game_audio.SoundHandle(99999))
        print("FAIL - No exception raised")
        audio.shutdown()
        return False
    except game_audio.InvalidHandleException:
        print("PASS")
        audio.shutdown()
        return True
    except Exception as e:
        print(f"FAIL - Wrong exception type: {type(e).__name__}: {e}")
        audio.shutdown()
        return False

def test_invalid_group_handle():
    """Test: Using invalid group handle should raise InvalidHandleException"""
    print("Test: Invalid group handle... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    audio.initialize()
    
    try:
        audio.set_group_volume(game_audio.GroupHandle(99999), 0.5)
        print("FAIL - No exception raised")
        audio.shutdown()
        return False
    except game_audio.InvalidHandleException:
        print("PASS")
        audio.shutdown()
        return True
    except Exception as e:
        print(f"FAIL - Wrong exception type: {type(e).__name__}: {e}")
        audio.shutdown()
        return False

def test_file_not_found():
    """Test: Loading non-existent file should raise FileLoadException"""
    print("Test: File not found... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    audio.initialize()
    
    try:
        audio.load_sound("nonexistent_file.mp3")
        print("FAIL - No exception raised")
        audio.shutdown()
        return False
    except game_audio.FileLoadException:
        print("PASS")
        audio.shutdown()
        return True
    except Exception as e:
        print(f"FAIL - Wrong exception type: {type(e).__name__}: {e}")
        audio.shutdown()
        return False

def test_sound_playback_initialization_failure():
    """Test: Sound playback initialization failure handling"""
    print("Test: Sound playback initialization failure... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    audio.initialize()
    
    try:
        # Test that valid sounds play correctly (regression test)
        # This verifies our fix doesn't break normal operation
        if sound_exists("digital_base.wav"):
            sound = audio.load_sound(get_sound_path("digital_base.wav"))
            audio.play_sound(sound)
            wait_ms(50)
            audio.stop_sound(sound)
            audio.destroy_sound(sound)
            
            # Verify exception hierarchy - FileLoadException should be catchable as AudioException
            # Note: In practice, ma_sound_init_from_file() can fail if:
            # - The file was deleted after loading but before playing
            # - The file is corrupted or in an unsupported format
            # - Audio device issues occur
            # - Memory allocation fails
            # If such a failure occurs, PlaySound() will throw FileLoadException
            
            # Test that FileLoadException can be caught as AudioException (hierarchy test)
            sound2 = audio.load_sound(get_sound_path("digital_base.wav"))
            try:
                audio.play_sound(sound2)
                wait_ms(50)
                audio.stop_sound(sound2)
                audio.destroy_sound(sound2)
                
                # If we get here, playback succeeded (expected for valid files)
                print("PASS")
                audio.shutdown()
                return True
            except game_audio.FileLoadException as e:
                # If FileLoadException is thrown, verify it has useful information
                msg = str(e)
                if "file" in msg.lower() or "playback" in msg.lower() or "initialize" in msg.lower():
                    print("PASS (exception with descriptive message)")
                    audio.destroy_sound(sound2)
                    audio.shutdown()
                    return True
                else:
                    print(f"FAIL - Exception message not descriptive: {msg}")
                    audio.destroy_sound(sound2)
                    audio.shutdown()
                    return False
            except game_audio.AudioException:
                # FileLoadException should be catchable as AudioException
                print("PASS (exception caught as AudioException)")
                audio.destroy_sound(sound2)
                audio.shutdown()
                return True
        else:
            print("SKIP - Test sound file not found")
            audio.shutdown()
            return True
    except Exception as e:
        print(f"FAIL - Unexpected exception: {type(e).__name__}: {e}")
        audio.shutdown()
        return False

def test_fade_duration_validation():
    """Test: Fade duration validation"""
    print("Test: Fade duration validation... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    audio.initialize()
    
    group = audio.create_group()
    
    # Test negative duration
    try:
        audio.fade_group(group, 0.5, timedelta(milliseconds=-100))
        print("FAIL - No exception for negative duration")
        audio.destroy_group(group)
        audio.shutdown()
        return False
    except game_audio.AudioException:
        pass  # Expected
    except Exception as e:
        print(f"FAIL - Wrong exception type: {type(e).__name__}: {e}")
        audio.destroy_group(group)
        audio.shutdown()
        return False
    
    # Test zero duration
    try:
        audio.fade_group(group, 0.5, timedelta(milliseconds=0))
        print("FAIL - No exception for zero duration")
        audio.destroy_group(group)
        audio.shutdown()
        return False
    except game_audio.AudioException:
        pass  # Expected
    except Exception as e:
        print(f"FAIL - Wrong exception type: {type(e).__name__}: {e}")
        audio.destroy_group(group)
        audio.shutdown()
        return False
    
    audio.destroy_group(group)
    audio.shutdown()
    print("PASS")
    return True

def test_input_validation():
    """Test: Input validation for layer names and paths"""
    print("Test: Input validation... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    audio.initialize()
    
    track = audio.create_track()
    
    # Test empty layer name
    try:
        if sound_exists("digital_base.wav"):
            audio.add_layer(track, "", get_sound_path("digital_base.wav"))
            print("FAIL - No exception for empty layer name")
            audio.destroy_track(track)
            audio.shutdown()
            return False
    except game_audio.AudioException:
        pass  # Expected
    except Exception as e:
        print(f"FAIL - Wrong exception type: {type(e).__name__}: {e}")
        audio.destroy_track(track)
        audio.shutdown()
        return False
    
    # Test empty filepath
    try:
        audio.add_layer(track, "layer1", "")
        print("FAIL - No exception for empty filepath")
        audio.destroy_track(track)
        audio.shutdown()
        return False
    except game_audio.AudioException:
        pass  # Expected
    except Exception as e:
        print(f"FAIL - Wrong exception type: {type(e).__name__}: {e}")
        audio.destroy_track(track)
        audio.shutdown()
        return False
    
    # Test empty folder path
    try:
        audio.play_random_sound_from_folder("", game_audio.GroupHandle(0))
        print("FAIL - No exception for empty folder path")
        audio.destroy_track(track)
        audio.shutdown()
        return False
    except game_audio.AudioException:
        pass  # Expected
    except Exception as e:
        print(f"FAIL - Wrong exception type: {type(e).__name__}: {e}")
        audio.destroy_track(track)
        audio.shutdown()
        return False
    
    audio.destroy_track(track)
    audio.shutdown()
    print("PASS")
    return True

def test_exception_hierarchy():
    """Test: Exception type hierarchy"""
    print("Test: Exception hierarchy... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    audio.initialize()
    
    # Test that specific exceptions can be caught as AudioException
    try:
        audio.play_track(game_audio.TrackHandle(0))
        print("FAIL - Expected exception to be thrown")
        audio.shutdown()
        return False
    except game_audio.AudioException:
        print("PASS")
        audio.shutdown()
        return True
    except Exception as e:
        print(f"FAIL - Wrong exception type: {type(e).__name__}: {e}")
        audio.shutdown()
        return False

def run_all_tests():
    """Run all validation tests"""
    print("=== Python Audio Validation Tests ===\n")
    
    tests = [
        test_not_initialized,
        test_audio_session_lifecycle,
        test_invalid_track_handle,
        test_invalid_sound_handle,
        test_invalid_group_handle,
        test_file_not_found,
        test_sound_playback_initialization_failure,
        test_fade_duration_validation,
        test_input_validation,
        test_exception_hierarchy,
    ]
    
    passed = 0
    failed = 0
    
    for test in tests:
        try:
            if test():
                passed += 1
            else:
                failed += 1
        except Exception as e:
            print(f"ERROR - {e}")
            failed += 1
    
    print(f"\n=== Results: {passed} passed, {failed} failed ===")
    return failed == 0

if __name__ == "__main__":
    success = run_all_tests()
    sys.exit(0 if success else 1)
